runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -119,20 +124,42 @@ public Object getPayload(CorfuRuntime runtime) {
     @Override
     public synchronized void releaseBuffer() {
         if (serializedCache != null) {
-            serializedCache.release();
-            if (serializedCache.refCnt() == 0) {
+            serializedCache.buffer.release();
+            if (serializedCache.buffer.refCnt() == 0) {
                 serializedCache = null;
             }
         }
     }
 
     @Override
-    public synchronized void acquireBuffer() {
+    public synchronized void acquireBuffer(boolean metadata) {
         if (serializedCache == null) {
-            serializedCache = Unpooled.buffer();
-            doSerializeInternal(serializedCache);
+            acquireBufferInternal(metadata);
         } else {
-            serializedCache.retain();
+            if (metadata) {
+                serializedCache.buffer.resetReaderIndex();
+                serializedCache.buffer.writerIndex(serializedCache.metadataOffset);
```

Comment:
Why do you need to reset the reader index? shouldn't you assert that its actually zero?

When would you read the payload, then reseralize the metadata portion and then reread ?

Reply:
Because in case of retries, the same LogData is read into the netty buffer, which changes readerIndex, in the retries, we need to reset reader index.